### PR TITLE
[SPARK-29937][SQL] Make FileSourceScanExec class fields lazy

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Since JIRA SPARK-28346,PR [25111](https://github.com/apache/spark/pull/25111), QueryExecution will copy all node stage-by-stage. This make all node instance twice almost. So we should make all class fields lazy to avoid create more unexpected object.

### Why are the changes needed?

Avoid create more unexpected object.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Exists UT.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @viirya @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -325,8 +325,7 @@ case class FileSourceScanExec(
   }
 
   @transient
-  private val pushedDownFilters = dataFilters.flatMap(DataSourceStrategy.translateFilter)
-  logInfo(s"Pushed Filters: ${pushedDownFilters.mkString(",")}")
```
Comment (ulysses-you) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: It is confused the log will print twice.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: If the log is consistent, I think it should not be confused.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 2]
```diff
@@ -177,6 +177,8 @@ object FileSourceStrategy extends Strategy with Logging {
       // Partition keys are not available in the statistics of the files.
       val dataFilters =
         normalizedFiltersWithoutSubqueries.filter(_.references.intersect(partitionSet).isEmpty)
+      logInfo(s"Pushed Filters: " +
+        s"${dataFilters.flatMap(DataSourceStrategy.translateFilter).mkString(",")}")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: We will run more than once translateFilter.

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Yes, but if we want to print the pushed filter once, I can not find a better way.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: @ulysses-you . Please don't piggy-back independent item into your PR. This is totally irrelevant to your PR (`Make FileSourceScanExec class fields lazy`)

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Yes, it is a related change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: I think run translateFilter twice is worse than logging twice.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: Got it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala]: Before this pr, translateFilter is also run twice. This just reduce one logging.
How about pass pushedDownFilters in `case class FileSourceScanExec()` ?

**Comment by cloud-fan** [comment]
OK so plan copy is more expensive than we expect. How about we reduce the number of copies? What we really need is to copy the logical plan between analyzer and optimizer, other copies are not really needed, and are just there for consistency.

**Comment by ulysses-you** [comment]
Yeah, it is really expensive, many plan do init action during instance.
I think we just need to copy analyzed plan what is the base plan. And for other case, we should check and copy just like what AdaptiveExecution do before.
```
val logicalPlan = if (sparkSession.sessionState.conf.adaptiveExecutionEnabled) {
  optimizedPlan.clone() 
} else {
  optimizedPlan
}
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you, @ulysses-you , @cloud-fan , @viirya .

**Comment by ulysses-you** [comment]
Thanks for merging!

**Pull request closed without merging**